bindings/wasm/src/account/storage/traits.rs

```diff
@@ -0,0 +1,292 @@
+      .as_string()
+      .ok_or_else(|| AccountStorageError::SerializationError("Expected string".to_string()))?;
+    let public_key: PublicKey = decode_b58(&public_key)?.into();
+    Ok(public_key)
+  }
+
+  /// Deletes the keypair specified by `location`.
+  async fn key_del(&self, did: &IotaDID, location: &KeyLocation) -> AccountStorageResult<()> {
+    let promise: Promise = Promise::resolve(&self.key_del(did.clone().into(), location.clone().into()));
+    let result: JsValueResult = JsFuture::from(promise).await.into();
+    result.into()
+  }
+
+  /// Signs `data` with the private key at the specified `location`.
+  async fn key_sign(&self, did: &IotaDID, location: &KeyLocation, data: Vec<u8>) -> AccountStorageResult<Signature> {
+    let promise: Promise = Promise::resolve(&self.key_sign(did.clone().into(), location.clone().into(), data));
+    let result: JsValueResult = JsFuture::from(promise).await.into();
+    let js_value: JsValue = result.account_err()?;
+    let signature: WasmSignature = downcast_js_value(js_value, "Signature")?;
+    Ok(signature.into())
+  }
+
+  /// Returns `true` if a keypair exists at the specified `location`.
+  async fn key_exists(&self, did: &IotaDID, location: &KeyLocation) -> AccountStorageResult<bool> {
+    let promise: Promise = Promise::resolve(&self.key_exists(did.clone().into(), location.clone().into()));
+    let result: JsValueResult = JsFuture::from(promise).await.into();
+    result.into()
+  }
+
+  /// Returns the chain state of the identity specified by `did`.
+  async fn chain_state(&self, did: &IotaDID) -> AccountStorageResult<Option<ChainState>> {
+    let promise: Promise = Promise::resolve(&self.chain_state(did.clone().into()));
+    let result: JsValueResult = JsFuture::from(promise).await.into();
+    let js_value: JsValue = result.account_err()?;
+    if js_value.is_null() || js_value.is_undefined() {
+      return Ok(None);
+    }
+    let chain_state: WasmChainState = downcast_js_value(js_value, "ChainState")?;
+    Ok(Some(chain_state.into()))
+  }
+
+  /// Set the chain state of the identity specified by `did`.
+  async fn set_chain_state(&self, did: &IotaDID, chain_state: &ChainState) -> AccountStorageResult<()> {
+    let promise: Promise = Promise::resolve(&self.set_chain_state(did.clone().into(), chain_state.clone().into()));
+    let result: JsValueResult = JsFuture::from(promise).await.into();
+    result.into()
+  }
+
+  /// Returns the state of the identity specified by `did`.
+  async fn state(&self, did: &IotaDID) -> AccountStorageResult<Option<IdentityState>> {
+    let promise: Promise = Promise::resolve(&self.state(did.clone().into()));
+    let result: JsValueResult = JsFuture::from(promise).await.into();
+    let js_value: JsValue = result.account_err()?;
+    if js_value.is_null() || js_value.is_undefined() {
+      return Ok(None);
+    }
+    let state: WasmIdentityState = downcast_js_value(js_value, "IdentityState")?;
+    Ok(Some(state.into()))
```

Comment:
Q: is `downcast_js_value` just used as an optimisation? Strictly speaking we could just go through `serde` JSON deserialization too?

Reply:
It's nice to remember this as an alternative for non-serializable structs, but I think all structs still need to be serializable to cross the FFI boundary between Wasm and Napi.

Reply:
Looked into a past conversation, and the actual reason for introducing the downcast function was `WasmDIDLease`, which could not be serialized. For all other types, technically we wouldn't need it. So it could be removed.
For `IdentityState` - because it contains an `IotaDocument` - not having to do the extra serialization is certainly a nice performance improvement, especially for larger documents.

I'm not sure how much of an attack surface this function is. It certainly feels safer to go down the serialization route, but is it worth the performance impact?